yi-core/test/Spec.hs

```diff
@@ -0,0 +1,49 @@
+import Test.Tasty
+import Test.Tasty.Hspec
+import Test.Tasty.QuickCheck
+
+import qualified Yi.CompletionTree as CT
+import           Data.List         (nub, sort)
+import qualified Data.Map          as M
+
+main :: IO ()
+main = defaultMain . tests =<< specTests
+
+tests :: TestTree -> TestTree
+tests specTests' = testGroup "CompletionTree Tests" [properties, specTests']
+
+properties :: TestTree
+properties = testGroup "(checked by QuickCheck)"
+  [ testProperty "CT.toList . CT.fromList == nub . sort" $
+      \list -> CT.toList (CT.fromList list) == sort (nub (list :: [String]))
+  , testProperty "update (fromList [a]) a == fromList [mempty] (a is a non-empty string)" $
+      \string -> null string || CT.update (CT.fromList [string :: String]) string == CT.fromList [mempty]
+  , testProperty "\"\" `elem` update (fromList [a,...]) a" $
+      \listOfStrings -> null listOfStrings || null (head listOfStrings) || "" `elem` CT.toList (CT.update (CT.fromList listOfStrings) (head listOfStrings))
+  , testProperty "complete (fromList [a]) == (a, fromList [\"\"])" $
+      \string -> CT.complete (CT.fromList [string]) == (string,CT.fromList [""])
+  ]
+
+specTests :: IO TestTree
+specTests = testSpec "Spec tests" $
```

Comment:
Does this return `IO` because of autodiscovery or something? This is honestly hideous but seems to me more like a limitation of hspec than your fault